### PR TITLE
Extension Resources

### PR DESCRIPTION
**Pull request by markcowl**
Note that this also:
- Removes unnecessary type arguments from Operations classes and interfaces
- Refactors ArmResource to be parallel to other resources
- Makes Operations interfaces naming consistent

### CODE REVIEW

File: azure-proto-authorization/AuthorizationOperations.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
+﻿using Azure.ResourceManager.Core;
+using System;
+
+namespace azure_proto_authorization
+{
+    public class AuthorizationOperations : OperationsBase
+    {
+        public AuthorizationOperations(ArmResourceOperations genericOperations)
```
Comment (m-nash) [inline azure-proto-authorization/AuthorizationOperations.cs:8]: should be internal constructor

File: azure-proto-authorization/AuthorizationOperations.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+            : this(genericOperations.ClientContext, genericOperations.Id, genericOperations.ClientOptions)
+        {
+        }
+
+        public AuthorizationOperations(ArmClientContext context, ResourceIdentifier id, ArmClientOptions options = default)
```
Comment (m-nash) [inline azure-proto-authorization/AuthorizationOperations.cs:13]: should be internal constructor

File: azure-proto-authorization/AuthorizationOperations.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+            : base(context, id, options)
+        {
+        }
+
+        public AuthorizationOperations(ArmClientContext context, Resource resource, ArmClientOptions options = default)
```
Comment (m-nash) [inline azure-proto-authorization/AuthorizationOperations.cs:18]: should be internal constructor

File: azure-proto-authorization/RoleAssignmentContainer.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+using System.Threading.Tasks;
+
+namespace azure_proto_authorization
+{
+    /// <summary>
+    /// Container for role assignments - note that in this case, the container is either a TrackedResource or a resource Id
+    /// </summary>
+    public class RoleAssignmentContainer : ExtensionResourceContainer<RoleAssignment, RoleAssignmentCreateParameters>
+    {
+        public RoleAssignmentContainer(OperationsBase operations) : base(operations)
```
Comment (m-nash) [inline azure-proto-authorization/RoleAssignmentContainer.cs:14]: should be internal constructor

File: azure-proto-authorization/RoleAssignmentContainer.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        {
+        }
+
+        public RoleAssignmentContainer(ArmClientContext context, TrackedResource scope) : base(context, scope) { }
```
Comment (m-nash) [inline azure-proto-authorization/RoleAssignmentContainer.cs:18]: should be internal constructor

File: azure-proto-authorization/RoleAssignmentContainer.cs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        public RoleAssignmentContainer(ArmClientContext context, ResourceIdentifier scope) : base(context, scope) { }
```
Comment (m-nash) [inline azure-proto-authorization/RoleAssignmentContainer.cs:19]: should be internal constructor

File: azure-proto-authorization/RoleAssignmentOperations.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+using Azure.ResourceManager.Authorization.Models;
+using Azure.ResourceManager.Core;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace azure_proto_authorization
+{
+    public class RoleAssignmentOperations : ExtensionResourceOperationsBase<RoleAssignment>, IDeletableResource
+    {
+        public RoleAssignmentOperations(ArmResourceOperations genericOperations) : this(genericOperations.ClientContext, genericOperations.Id) { }
```
Comment (m-nash) [inline azure-proto-authorization/RoleAssignmentOperations.cs:12]: should be internal constructor

File: azure-proto-authorization/RoleAssignmentOperations.cs [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+        public RoleAssignmentOperations(ArmClientContext context, ResourceIdentifier id) : this(context, new ArmResourceData(id)) { }
```
Comment (m-nash) [inline azure-proto-authorization/RoleAssignmentOperations.cs:14]: should be internal constructor

File: azure-proto-authorization/RoleAssignmentOperations.cs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+        public RoleAssignmentOperations(ArmClientContext context, Resource resource) : base(context, resource)
```
Comment (m-nash) [inline azure-proto-authorization/RoleAssignmentOperations.cs:16]: should be internal constructor

File: lib/Azure.ResourceManager.Authorization.deps.json [hunk 10]
```diff
@@ -0,0 +1,347 @@
+{
```
Comment (m-nash) [inline lib/Azure.ResourceManager.Authorization.deps.json:1]: should this be a nuget dependency vs a local binary?  That is what we have done with the others

**Review by m-nash — CHANGES_REQUESTED** [review]

File: Azure.ResourceManager.Core.Tests/ResourceIdentifierTests.cs [hunk 11]
```diff
@@ -25,6 +37,22 @@ public void CanParseRPIds(string subscription, string resourceGroup, string prov
...
+        [TestCase(SubscriptionResourceId, "Microsoft.Authorization", "roleAssignments", "myRa")]
+        public void CanParseExtensionResourceIds(string baseId, string extensionNamespace, string extensionType, string extensionName)
+        {
+            ResourceIdentifier targetResourceId = baseId;
+            ResourceIdentifier subject = $"{baseId}/providers/{extensionNamespace}/{extensionType}/{extensionName}";
+            ResourceType expectedType = $"{extensionNamespace}/{extensionType}";
+            Assert.AreNotEqual(targetResourceId.Type, subject.Type);
+            Assert.AreEqual(expectedType, subject.Type);
+            Assert.NotNull(subject.Parent);
+            Assert.AreEqual(targetResourceId, subject.Parent);
```
[the remark this answers is not in the document]
Comment (markcowl) [reply Azure.ResourceManager.Core.Tests/ResourceIdentifierTests.cs:53]: I think this properly belongs in the test above, astargetResourceId is just one of the other kinds of resource ids we are already testing.

File: Azure.ResourceManager.Core/ExtensionResourceOperationsBase.cs [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+        /// Initializes a new instance of the <see cref="ExtensionResourceOperationsBase{TOperations, TResource}"/> class.
+        /// </summary>
+        /// <param name="context">Client configuration properties for these operations</param>
+        /// <param name="resource">The extention resource for operatiosn to act upon</param>
+        /// <param name="options">The client options to sue with these operations.</param>
+        public ExtensionResourceOperationsBase(ArmClientContext context, Resource resource, ArmClientOptions options = default)
+            : this(context, resource.Id, options)
+        {
+        }
+
```
[the remark this answers is not in the document]
Comment (markcowl) [reply Azure.ResourceManager.Core/ExtensionResourceOperationsBase.cs:99]: I think I got this, no stylecop warnings in this file.

File: lib/Azure.ResourceManager.Authorization.deps.json (hunk 10 above)
Comment (markcowl) [reply lib/Azure.ResourceManager.Authorization.deps.json:1]: This package isn't released, so it cannot be a nuget dependency.

File: azure-proto-authorization/Extensions/RoleAssignmentExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+
+        public static RoleAssignmentContainer RoleAssigmentsAtScope(this SubscriptionOperations resource, ResourceIdentifier scope)
+        {
+            return new RoleAssignmentContainer(resource.ClientContext, scope);
+        }
+        public static RoleAssignmentContainer RoleAssigmentsAtScope(this SubscriptionOperations resource, Resource scope)
+        {
+            return new RoleAssignmentContainer(resource.ClientContext, scope.Id);
+        }
+
```
[the remark this answers is not in the document]
Comment (markcowl) [reply azure-proto-authorization/Extensions/RoleAssignmentExtensions.cs:34]: These are all cleaned up

**Review by m-nash — APPROVED** [review]

**Pull request merged**